vm/proxyapp/proxyappclient.go

```diff
@@ -0,0 +1,494 @@
+// Copyright 2022 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+// Package proxyapp package implements the experimental plugins support.
+// We promise interface part will not be stable until documented.
+package proxyapp
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"math"
+	"net/rpc"
+	"net/rpc/jsonrpc"
+	"os"
+	"os/exec"
+	"sync"
+	"time"
+
+	"github.com/google/syzkaller/pkg/config"
+	"github.com/google/syzkaller/pkg/log"
+	"github.com/google/syzkaller/pkg/osutil"
+	"github.com/google/syzkaller/pkg/report"
+	"github.com/google/syzkaller/vm/proxyapp/proxyrpc"
+	"github.com/google/syzkaller/vm/vmimpl"
+)
+
+func osutilCommand(bin string, args ...string) subProcessCmd {
+	return osutil.Command(bin, args...)
+}
+
+type subProcessCmd interface {
+	StdinPipe() (io.WriteCloser, error)
+	StdoutPipe() (io.ReadCloser, error)
+	StderrPipe() (io.ReadCloser, error)
+	Start() error
+	Wait() error
+}
+
+func makeDefaultParams() *proxyAppParams {
+	return &proxyAppParams{
+		CommandRunner:   osutilCommand,
+		MakePipe:        osutil.LongPipe,
+		InitRetryDelay:  10 * time.Second,
+		MaxInitAttempts: math.MaxInt64,
+	}
+}
+
+func init() {
+	vmimpl.Register(
+		"proxyapp",
+		func(env *vmimpl.Env) (vmimpl.Pool, error) {
+			return ctor(makeDefaultParams(), env)
+		},
+		false)
+}
+
+type pool struct {
+	mu              sync.Mutex
+	env             *vmimpl.Env
+	proxy           *ProxyApp
+	ctxRunner       context.Context
+	ctxRunnerCancel context.CancelFunc
+	count           int
+}
+
+func (pool *pool) reset() {
+	pool.cancelRun()
+	pool.ctxRunner, pool.ctxRunnerCancel = context.WithCancel(context.TODO())
+	pool.proxy = nil
+	pool.count = 0
```

Comment:
I don't think manager code is prepared for pool size changes. It quires it once on start and then assumes it stays constant. So it's better to check that the app keeps the size consistent as well.

Reply:
Fixed. No test covers it.